### PR TITLE
fix: add a artifact_uri argument and config

### LINKED ISSUE #86: Properly test the use of mlflow in the scripts with the functional tests
### Problem

The functional tests only launch some scripts with the --mlflow_experiment argument to see if they run successfully or not. They should also check that the right items are saved in mlflow

## Concerned template

  - [x] NLP template
  - [x] NUM template
  - [ ] VISION template
  - [ ] How templates are generated - Jinja

### Solution

In the functional tests, add an explicit check for the local mlflow files

### LINKED ISSUE #87: No artifact_uri specified when generating a project
### Describe the bug

No artifact_uri is specified when generating a project. The artifact are saved where the script is run which can result in artifacts not appearing in the mlflow UI

## Concerned template

  - [x] NLP template
  - [x] NUM template
  - [ ] VISION template
  - [ ] How templates are generated - Jinja

### To Reproduce

Code snippet: 

```
text = 'coucou'
logger = MLflowLogger(experiment_name='coucou', tracking_uri='http://127.0.0.1:5000')
logger.log_metric('my_metric', 3453)
logger.log_text(text, 'text.txt')
logger.end_run()

```

1. Open mlflow ui (let's assume that the uri indicated is http://127.0.0.1:5000)
2. Execute the code snippet above somewhere else
3. Check mlflow UI,. The metric is properly saved but the artifact has been saved somewhere else and don't appear in the UI

### Expected behavior

The artifact should appear in mlflow UI by being saved at the proper place by specifying an artifact_uri

### Actual behavior

The artifacts are not saved at the right place. 

### Is this a regression?

That is, did this use to work the way you expected in the past?

- [ ] Yes
- [x] No

### Debug info

- Gabarit version: 1.2.5-dev
- Python version: (get it with `$ python --version`) 3.8
- OS version: ubuntu 20.04

### PR DESCRIPTION
**Pull request by h4c5**
## ✒️ Context

No artifact_uri is specified when generating a project. The artifact are saved where the script is run which can result in artifacts not appearing in the mlflow UI

- What kind of change does this PR introduce ?

  - [x] Bugfix
  - [ ] Feature
  - [ ] Refactoring
  - [ ] Other, please describe:

## 🧱 Description of Changes

- Add a artifact_uri configuration parameter
- By default store all artifacts in package_name-data/experiments/mlruns_artifacts 
- Add functional tests for scripts that use mlflow logging

  - [x] This is a change for the NLP template
  - [x] This is a change for the NUM template
  - [x] This is a change for the VISION template
  - [ ] This is a change for the API template
  - [ ] This changes how templates are generated (i.e. a Jinja change)

## 🩺 Testing

- Local test

  - [ ] This change does not need new tests
  - [x] Added/Updated unit tests
  - [x] Added/Updated functionals tests (i.e. e2e)

## 🔗 References

- **Issue**: Closes #86
- **Issue**: Closes #87
---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the GNU AFFERO GENERAL PUBLIC LICENSE.

### CODE REVIEW

File: gabarit/template_nlp/nlp_project/package_name/monitoring/mlflow_logger.py [hunk 1]
```diff
@@ -36,28 +37,48 @@
...
+            experiment = mlflow.get_experiment_by_name(experiment_name)
         except Exception as e:
             self.logger.error(repr(e))
             raise ConnectionError(f"Can't reach MLflow at {self.tracking_uri}. Please check the URI.")
+
+        # If the experiment exists, we recover experiment id and artifact_uri (which is link to the experiment)
+        if experiment:
+            experiment_id = experiment.experiment_id
+            artifact_uri = experiment.artifact_location
+        # Otherwise we create a new experiment with the provided artifact_rui
```
Comment (LexABzH) [inline gabarit/template_nlp/nlp_project/package_name/monitoring/mlflow_logger.py:72]: artifact_rui -> artifact_uri

File: gabarit/template_nlp/nlp_project/package_name/monitoring/mlflow_logger.py [hunk 2]
```diff
@@ -36,28 +37,48 @@
...
+        else:
+            experiment_id = mlflow.create_experiment(experiment_name, artifact_location=artifact_uri)
+        
+        mlflow.set_experiment(experiment_id=experiment_id)
+
+        self._experiment_id = experiment_id
+        self._experiment_name = experiment_name
+        self._artifact_uri = artifact_uri
+
```
Comment (LexABzH) [inline gabarit/template_nlp/nlp_project/package_name/monitoring/mlflow_logger.py:81]: Do we really need properties here ? (no setter + does nothing special)

File: gabarit/template_nlp/nlp_project/package_name/monitoring/mlflow_logger.py [hunk 3]
```diff
@@ -70,6 +91,21 @@ def tracking_uri(self, uri:str) -> None:
...
+    @property
+    def experiment_name(self) -> str:
+        '''Experiment name. It can not be changed.'''
+        return self._experiment_name
+
+    @property
+    def artifact_uri(self) -> str:
+        '''Experiment artifact URI. It can not be changed.'''
+        return self._artifact_uri
+
```
Comment (LexABzH) [inline gabarit/template_nlp/nlp_project/package_name/monitoring/mlflow_logger.py:108]: See above comment

File: gabarit/template_nlp/nlp_project/package_name/monitoring/mlflow_logger.py [hunk 4]
```diff
@@ -36,28 +37,48 @@
...
         # Set tracking URI & experiment name
         self.tracking_uri = tracking_uri
-        # No need to set_tracking_uri, this is done through the setter decorator
-        self.experiment_name = experiment_name
+        
+        # Get the experiment if it exists and check if there is a connection error by doing it
         try:
-            mlflow.set_experiment(experiment_name)
+            experiment = mlflow.get_experiment_by_name(experiment_name)
```
Comment (LexABzH) [inline gabarit/template_nlp/nlp_project/package_name/monitoring/mlflow_logger.py:63]: Does that return an empty value if no experiment exists with this name ?

File: gabarit/template_nlp/nlp_project/package_name/monitoring/mlflow_logger.py (hunk 2 above)
Comment (h4c5) [reply gabarit/template_nlp/nlp_project/package_name/monitoring/mlflow_logger.py:81]: In combination with the @property decorator it helps to protect these attributes from being set.

File: gabarit/template_nlp/nlp_project/package_name/monitoring/mlflow_logger.py (hunk 3 above)
Comment (h4c5) [reply gabarit/template_nlp/nlp_project/package_name/monitoring/mlflow_logger.py:108]: In combination with the @property decorator it helps to protect these attributes from being set.

File: gabarit/template_nlp/nlp_project/package_name/monitoring/mlflow_logger.py (hunk 4 above)
Comment (h4c5) [reply gabarit/template_nlp/nlp_project/package_name/monitoring/mlflow_logger.py:63]: Yes, it returns `None`.

**Review by LexABzH — APPROVED** [review]

**Pull request merged**